utilities/src/bits.rs

```diff
@@ -126,11 +142,27 @@ macro_rules! impl_bits_for_integer {
                 bits_le
             }
 
+            #[inline]
+            fn write_bits_le(&self, vec: &mut Vec<bool>) {
+                vec.reserve(<$int>::BITS as usize);
+                let mut value = *self;
+                for _ in 0..<$int>::BITS {
+                    vec.push(value & 1 == 1);
+                    value = value.wrapping_shr(1u32);
+                }
+            }
+
             /// Returns `self` as a boolean array in big-endian order.
             #[inline]
             fn to_bits_be(&self) -> Vec<bool> {
```

Comment:
Is there a way (without perf penalty) to have the `to_bits_le/be` use the `write_bits_le/be` impl?

Reply:
yes; it's basically what https://github.com/AleoHQ/snarkVM/pull/1836 did (just with an additional trait); with a single trait the trick would be to have a default implementation of `ToBits::to_bits_le/be` (instead of `::write_bits_le/be` like in the current shape of this PR) implemented in terms of `::write_bits_le/be` and then to reimplement every existing instance of `ToBits::to_bits_le/be` so that they also use `::write_bits_le/be` internally; that way, every top-level/outermost call to `ToBits::to_bits_le` would explicitly allocate only once, as every inner call would extend an initially created vector.

Reply:
note on the aforementioned perf penalty: this PR on its own also doesn't introduce any, because there are currently no other users of `ToBits::write_bits_le/be` and the ones that do use it reimplement it in a performant fashion.

Reply:
this is how the trait would look like:
```
diff --git a/utilities/src/bits.rs b/utilities/src/bits.rs
index 64478054d..10fc29b00 100644
--- a/utilities/src/bits.rs
+++ b/utilities/src/bits.rs
@@ -18,26 +18,26 @@ use anyhow::Result;
 
 pub trait ToBits: Sized {
     /// Returns `self` as a boolean array in little-endian order.
-    fn to_bits_le(&self) -> Vec<bool>;
+    fn to_bits_le(&self) -> Vec<bool> {
+        let mut bits = vec![];
+        self.write_bits_le(&mut bits);
+        bits
+    }
 
     /// Writes `self` into the given vector as a boolean array
     /// in little-endian order.
-    fn write_bits_le(&self, vec: &mut Vec<bool>) {
-        // A default, unoptimized implementation.
-        let bits = self.to_bits_le();
-        vec.extend_from_slice(&bits);
-    }
+    fn write_bits_le(&self, vec: &mut Vec<bool>);
 
     /// Returns `self` as a boolean array in big-endian order.
-    fn to_bits_be(&self) -> Vec<bool>;
+    fn to_bits_be(&self) -> Vec<bool> {
+        let mut bits = vec![];
+        self.write_bits_be(&mut bits);
+        bits
+    }
 
     /// Writes `self` into the given vector as a boolean array
     /// in big-endian order.
-    fn write_bits_be(&self, vec: &mut Vec<bool>) {
-        // A default, unoptimized implementation.
-        let bits = self.to_bits_be();
-        vec.extend_from_slice(&bits);
-    }
+    fn write_bits_be(&self, vec: &mut Vec<bool>);
 }
 
 pub trait FromBits: Sized {
```
and this is an example of how one of its implementations (the changes to others would be very similar) would change:
```
diff --git a/utilities/src/biginteger/bigint_256.rs b/utilities/src/biginteger/bigint_256.rs
index 188f68353..6f9754dfa 100644
--- a/utilities/src/biginteger/bigint_256.rs
+++ b/utilities/src/biginteger/bigint_256.rs
@@ -227,13 +227,13 @@ impl crate::biginteger::BigInteger for BigInteger256 {
 
 impl ToBits for BigInteger256 {
     #[doc = " Returns `self` as a boolean array in little-endian order, with trailing zeros."]
-    fn to_bits_le(&self) -> Vec<bool> {
-        BitIteratorLE::new(self).collect::<Vec<_>>()
+    fn write_bits_le(&self, vec: &mut Vec<bool>) {
+        vec.extend(BitIteratorLE::new(self));
     }
 
     #[doc = " Returns `self` as a boolean array in big-endian order, with leading zeros."]
-    fn to_bits_be(&self) -> Vec<bool> {
-        BitIteratorBE::new(self).collect::<Vec<_>>()
+    fn write_bits_be(&self, vec: &mut Vec<bool>) {
+        vec.extend(BitIteratorBE::new(self));
     }
 }

```